File: backend/internal/models/wiki_article.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"fmt"

	"github.com/lib/pq"
	"gorm.io/gorm"
)

type InfoboxData map[string]string

func (i InfoboxData) Value() (driver.Value, error) {
	if i == nil {
		return nil, nil
	}
	b, err := json.Marshal(i)
	return string(b), err
}

func (i *InfoboxData) Scan(value interface{}) error {
	if value == nil {
		return nil
	}
	bytes, ok := value.([]byte)
	if !ok {
		return fmt.Errorf("InfoboxData: cannot scan type %T", value)
	}
	return json.Unmarshal(bytes, i)
}

type WikiArticle struct {
	gorm.Model
	Slug        string         `gorm:"uniqueIndex;not null" json:"slug"`
	Title       string         `gorm:"not null" json:"title"`
	Subtitle    string         `json:"subtitle"`
	Category    string         `json:"category"`
	LeadText    string         `json:"leadText"`
	BodyHTML    string         `json:"bodyHtml"`
	InfoboxData InfoboxData    `gorm:"type:jsonb" json:"infoboxData"`
	Tags        pq.StringArray `gorm:"type:text[]" json:"tags"`
	AuthorID    *uint          `json:"authorId"`
	Author      User           `json:"author,omitempty"`
	IsApproved  bool           `gorm:"default:false" json:"isApproved"`
	ViewCount   int            `gorm:"default:0" json:"viewCount"`

	Revisions []WikiRevision `gorm:"foreignKey:ArticleID" json:"revisions,omitempty"`
}

type WikiRevision struct {
	gorm.Model
	ArticleID uint   `gorm:"not null" json:"articleId"`
	AuthorID  uint   `gorm:"not null" json:"authorId"`
	Author    User   `json:"author,omitempty"`
	BodyHTML  string `json:"bodyHtml"`
	Summary   string `json:"summary"`
}

File: mobile/lib/models/wiki_article.dart
class WikiArticle {
  final int id;
  final String title;
  final String slug;
  final String bodyHtml;
  final String summary;
  final String category;
  final String imageUrl;
  final String createdAt;
  final String updatedAt;

  const WikiArticle({
    required this.id,
    required this.title,
    this.slug = '',
    this.bodyHtml = '',
    this.summary = '',
    this.category = '',
    this.imageUrl = '',
    this.createdAt = '',
    this.updatedAt = '',
  });

  factory WikiArticle.fromJson(Map<String, dynamic> json) => WikiArticle(
        id: json['id'] ?? json['ID'] ?? 0,
        title: json['title'] ?? '',
        slug: json['slug'] ?? '',
        bodyHtml: json['bodyHtml'] ?? json['body_html'] ?? '',
        summary: json['summary'] ?? '',
        category: json['category'] ?? '',
        imageUrl: json['imageUrl'] ?? '',
        createdAt: json['createdAt'] ?? json['CreatedAt'] ?? '',
        updatedAt: json['updatedAt'] ?? json['UpdatedAt'] ?? '',
      );
}
